### PR TITLE
Switch to Generic model interface

### PR DESCRIPTION
**Pull request by jkulhanek**
This PR changes the Model's class interface to accept the type of the config. This should improve the type hints for the models (as you no longer need to override the __init__ method).
But I am not sure if this is needed or if using Generic could break compatibility for older python versions.
Please feel free to reject this PR if you are not sure of the benefits.

### CODE REVIEW

File: docs/developer_guides/config.md [hunk 1]
```diff
@@ -31,14 +31,12 @@ class NerfactoModelConfig(ModelConfig):
     _target: Type = field(default_factory=lambda: NerfactoModel)
     ...
 
-class NerfactoModel(Model):
+class NerfactoModel(Model[NerfactoModelConfig]):
      """Nerfacto model
 
     Args:
         config: Nerfacto configuration to instantiate model
     """
```
Comment (tancik) [inline docs/developer_guides/config.md:39]: Remove

File: docs/developer_guides/pipelines/models.md [hunk 2]
```diff
@@ -111,15 +111,13 @@ There are a lot of options! Thankfully, our config system makes this easy to han
 Furthermore, you have Python autocomplete and static checking working in your favor. At the top of every Model, we specify the config and then can easily pull of values throughout the implementation. Let's take a look at the beginning of the NerfactoModel implementation.
 
 ```python
-class NerfactoModel(Model):
+class NerfactoModel(Model[NerfactoModelConfig]):
     """Nerfacto model
 
     Args:
         config: Nerfacto configuration to instantiate model
     """
```
Comment (tancik) [inline docs/developer_guides/pipelines/models.md:119]: Remove here and in all other models since it is no longer an argument.

**Review by tancik — COMMENTED** [review]
This is a great idea! So much cleaner!

**Comment by jkulhanek** [comment]
Now I am not sure if it is a good idea. The api should be consistent, so we would want to apply this also to DataManager and Pipeline classes. The problem is for example in DynamicBatchPipeline. This class inherits from VanillaPipeline with a fixed config type, but it changes the config type on top. I am afraid that implementing this would make the API less consistent.

**Comment by jkulhanek** [comment]
I suggest to close this for now. What do you think?

**Comment by tancik** [comment]
I suppose. I do really like how this makes the code cleaner. I wonder if there are ways we can refactor the `DynamicBatchPipeline` so that it can fit into this framework.

**Comment by jkulhanek** [comment]
This will be elegant in Python 3.12 (https://peps.python.org/pep-0696/). I will try to think of something for the currently supported Python.

**Comment by jkulhanek** [comment]
It may be part of typing_extensions ...

**Comment by jkulhanek** [comment]
What do you think now?

File: nerfstudio/data/datamanagers/base_datamanager.py [hunk 3]
```diff
@@ -352,9 +346,6 @@ def _get_pixel_sampler(  # pylint: disable=no-self-use
         self, dataset: InputDataset, *args: Any, **kwargs: Any
     ) -> PixelSampler:
         """Infer pixel sampler to use."""
-        if self.config.patch_size > 1:
-            return PatchPixelSampler(*args, **kwargs, patch_size=self.config.patch_size)
-
```
Comment (tancik) [inline nerfstudio/data/datamanagers/base_datamanager.py:357]: Merge issue?

Comment (jkulhanek) [reply nerfstudio/data/datamanagers/base_datamanager.py]: Yes! Sorry for that.

**Comment by tancik** [comment]
It looks good to me. I'd like to get @brentyi opinion.

**Review by brentyi — COMMENTED** [review]
+1 for looks good, big fan of PRs with more lines deleted than added. There's always more to do but I'd support merging. :)

Only important thing is that this blurb in `docs/developer_guides/config.md` probably needs a tweak:
```
You can then enable type/auto complete on the config passed into the `NerfactoModel` by specifying the config type below the class definition.
```

A pair of thoughts related to this PR, that I don't feel strongly about:
- It would be nice if `.instantiate()` in `InstantiateConfig` was typed correctly. I think this is doable with another generic + some special handling for forward references (since configs for a model are defined before the model). This is a lot of generics but we could get rid of the `_target` fields that are scattered everywhere...
```python
@dataclass
class SomeConfig(InstantiateConfig[Model]):
    ...

class SomeModel(Model[SomeConfig]):
    ...
```
- Alternatively, if we generalize this generic structure to a `Configurable` class that inherits from `Generic[TConfig]`, we can assign `_target` values to config dataclasses automatically/dynamically by overriding `__class_getitem__` in `Configurable`.

File: nerfstudio/utils/misc.py [hunk 4]
```diff
@@ -20,9 +20,12 @@
 from typing import Any, Callable, Dict, List, Optional, Union
 
 import torch
+from typing_extensions import TypeVar
 
+T = TypeVar("T")
```
[the remark this answers is not in the document]
Comment (jkulhanek) [reply nerfstudio/utils/misc.py:25]: I had this first, but it failed because, in some cases, T is also `int`.

**Comment by jkulhanek** [comment]
> * Alternatively, if we generalize this generic structure to a `Configurable` class that inherits from `Generic[TConfig]`, we can assign `_target` values to config dataclasses automatically/dynamically by overriding `__class_getitem__` in `Configurable`.

Adding this would introduce one-to-one correspondence between configs and models. This is not the case now, as multiple models share same config.

File: nerfstudio/utils/misc.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (brentyi) [reply nerfstudio/utils/misc.py:25]: Oh interesting, yeah I see, when this function says `stuff` it really means it huh

**Comment by brentyi** [comment]
> Adding this would introduce one-to-one correspondence between configs and models. This is not the case now, as multiple models share same config.

Is this true beyond mipnerf using the vanilla nerf config structure? This is the only one I'm aware of, and it'd be quick to add a dummy class. The base `InstantiateConfig` class could also still have a `_target` field that can be overridden as is currently done, it just wouldn't need to be repeated in every config class definition.

**Comment by jkulhanek** [comment]
Sure, I agree, I will do it.
________________________________

**Comment by jkulhanek** [comment]
Oh, another problem:
```
tyro._instantiators.UnsupportedTypeAnnotationError: Default values for generic subparsers are not supported.
```

**Comment by brentyi** [comment]
Shoot! Are you able to share a fork / branch that causes this error? I can check whether this is supportable or not.

Alternatively, I think it would make sense to get this PR merged before conflicts start creeping in and we can revisit next steps in a separate PR. (if you have time)

**Comment by jkulhanek** [comment]
I have added a proof-of-concept implementation with forward refs (correct types for instantiated objects). Currently, it fails with the error I mentioned. But what do you think of the overall concept?

**Comment by jkulhanek** [comment]
I don’t think that there is a way to make correct type annotations for both the config and the model without forward references. Personally, I would be ok with only having correct annotations for the model and not for the “setup” method of config (avoiding forward refs and using _target). I am also ok with the code in the current PR with forward refs-I believe I can get around the tyro bug. Which one would you prefer?

I also won’t have time to work on it till 9th of March
________________________________

**Comment by brentyi** [comment]
> Personally, I would be ok with only having correct annotations for the model and not for the “setup” method of config (avoiding forward refs and using _target).

Yeah, I also don't feel strongly, but have heard frustration expressed about verbose/confusing typing syntax so let's just go with this.

Maybe we could think about a bigger overhaul at some point, the way PyTorch modules are instantiated + structured makes it hard but the [Implicitron config](https://pytorch3d.org/tutorials/implicitron_config_system) system, [hydra-zen](https://github.com/mit-ll-responsible-ai/hydra-zen), what nerfstudio has etc all have subsets of nice usability things (conciseness, type safety, ease-of-use with inheritance + documentation, ...). It would be nice to get all of them, with less syntax and less tradeoffs...

**Comment by tancik** [comment]
Is this worth reviving, or should we close it for now?

**Comment by jkulhanek** [comment]
Let's just close it.

**Pull request closed without merging**